README.md
# sliver
I didn't see any existing Node solutions for Boundary's Flake or Twitter's Snowflake style ids that did everything we wanted as fast as we'd like. Here's our attempt at a clone that does the following:

 * 128 bit keys
 * K-ordered
 * Rendered as base 62
 * Lexicographically sortable
 * Coordination free-ish

### Break down
128 bit key comprised of the following:

`{timestamp}{worker}{sequence}`

 * 64 bit timestamp
 * 48 bit worker id
 * 16 bit sequence number

The sequence number increments for each subsequent id requested within the same millisecond.

## API

### Using MACAddress or HostName as seed
If no seed is provided, Sliver will use either the MAC address or host name plus the process id as the seed. This should prevent duplicate seeds and id collisions, but be aware of any scenario that might cause duplicate MACs or host names in your environment.

```javascript

// reading the MAC address is async, hence the ready call.
var sliver = require( 'sliver' )( ready );
id = sliver.getId(); // ta-da
```

### Unique seeds
Anything can seed the node id. sliver uses murmurhash3 to create a unique 32 bit integer from whatever you have lying around. This needs to be unique for every instance creating ids.

```javascript
// no need to provide a call back when specifying a seed
var sliver = require( 'sliver' )( 'Hey, look, a string based seed.' );
id = sliver.getId(); // ta-da
```

### Speed
Looks like this tops out around 20 / ms on modern processors. Do with that what you will :)


CHANGELOG.md
## Changelog

## 0.2.*

### 0.2.0
 * Update basejump version to support Node 5.
 * Switch to farmhash (faster and Node 5 friendly).

## 0.1.*

### 0.1.0

 * Update to basejump version that supports Node versions v0.10.* and v0.12.*
 * Update to use MAC or Machine and PID as default worker id to ensure uniqueness


spec/sliver.spec.js
require( 'should' );
var path = require( 'path' ),
	_ = require( 'lodash' );

describe( 'when starting up without a seed', function() {
	var sliver;
	before( function( done ) {
		sliver = require( '../src/sliver.js' )( done );
	} );

	it( 'should seed from MAC Address', function() {
		sliver.seed.length.should.equal( 6 );
	} );

	it( 'should start with instance count of 1', function() {
		sliver.index.should.equal( 1 );
	} );

	describe( 'if requiring a new instance', function() {
		before( function( done ) {
			sliver = require( '../src/sliver.js' )( done );
		} );

		it( 'should not create an additional instance', function() {
			sliver.index.should.equal( 1 );
		} );
	} );

	describe( 'when getting an id', function() {
		var id;
		before( function() {
			id = sliver.getId();
		} );

		it( 'should get a valid id', function() {
			id.length.should.equal( 22 );
		} );
	} );

	describe( 'when getting 10,000 ids', function() {
		var diff;
		var list = [];
		var idCount = 10000;
		this.timeout( 10000 );
		before( function( done ) {
			var start = process.hrtime();
			for (i = 0; i < idCount; i++) {
				list.push( sliver.getId() );
				if ( list.length == idCount ) {
					diff = process.hrtime( start );
					diff = ( diff[ 0 ] * 1e9 + diff[ 1 ] ) / 1000000;
					done();
				}
			}
		} );

		it( 'should produce only unique ids', function() {
			_.unique( list ).length.should.equal( idCount );
		} );

		it( 'should average 10 ids/ms', function() {
			( idCount / diff ).should.be.greaterThan( 10 );
		} );
	} );
} );

describe( 'when starting with a seed', function() {
	var sliver;
	before( function( done ) {
		var sliverPath = path.resolve( './src/sliver.js' );
		delete require.cache[ sliverPath ];
		sliver = require( '../src/sliver.js' )( 'burple', done );
	} );

	it( 'should seed from seed value', function() {
		sliver.seed.length.should.equal( 6 );
	} );

	it( 'should start with instance count of 1', function() {
		sliver.index.should.equal( 1 );
	} );
} );


src/sliver.js
var _ = require( 'lodash' );
var getMac = require( 'getmac' ).getMac;
var jump = require( 'basejump' );
var fh = require( 'farmhash' );
var generator;
var instanceCount = 0;

// 128 bit ids can generate up to
//          1    1    2    2    3    3   3
//     5    0    5    0    5    0    5   9
// 340282366920938463463374607431768211456
// unique ids.
// By base-62 encoding them, we can shorten the length of the id to 22 places

// via http://stackoverflow.com/questions/8482309/converting-javascript-integer-to-byte-array-and-back
function longToBytes( long ) {
	// we want to represent the input as a 8-bytes array
	var byteArray = [ 0, 0, 0, 0, 0, 0, 0, 0 ];
	for (var index = 0; index < byteArray.length; index++) {
		var byte = long & 0xff;
		byteArray [ index ] = byte;
		long = ( ( long - byte ) / 256 );
	}
	return byteArray.reverse();
}

function getWorkerIdFromEnvironment( done ) {
	getMac( function( err, address ) {
		var bytes, id;
		if ( err ) {
			address = os.hostname();
		}
		bytes = getNodeBytesFromSeed( address );
		done( bytes );
	} );
}

function getNodeBytesFromSeed( seed ) {
	var hash = fh.hash32( [ seed.toString(), process.pid ].join( '|' ) );
	bytes = longToBytes( hash ).splice( 2 );
	return bytes;
}

var Sliver = function( seed, ready ) {
	ready = ready || function() {};
	this.lastMs = Date.now();
	this.msCounter = 0;
	this.seed = seed;
	this.msBytes = [ 0, 0 ];
	this.index = ++instanceCount;
	if ( !seed ) {
		getWorkerIdFromEnvironment( function( bytes ) {
			this.seed = bytes;
			ready();
		}.bind( this ) );
	} else {
		this.seed = getNodeBytesFromSeed( seed );
		ready();
	}
};

Sliver.prototype.getId = function() {
	this.updateTime();
	var bytes = this.timestamp.concat( this.seed, this.msBytes ).reverse();
	return jump.toBase62( bytes, 22 );
};

Sliver.prototype.updateTime = function() {
	var now = Date.now();
	var change = false;
	if ( now < this.lastMs ) {
		throw new Error( 'Everything you know is a lie; NTP just jumped in space-time!' );
	} else if ( now > this.lastMs ) {
		change = true;
		this.msCounter = 0;
		this.msBytes = [ 0, 0 ];
		this.lastMs = now;
		this.timestamp = longToBytes( now );
	} else {
		this.msCounter++;
		this.msBytes = [ 0, 0 ].concat( longToBytes( this.msCounter ) ).splice( -2, 2 );
	}
	return change;
};

module.exports = function( seed, ready ) {
	if ( _.isFunction( seed ) ) {
		ready = seed;
		seed = undefined;
	}
	if ( generator && ready ) {
		ready();
	} else {
		generator = new Sliver( seed, ready );
	}
	return generator;
};
